src/claylib.lisp

```diff
@@ -39,19 +39,37 @@ background of the render texture, or NIL to skip clearing."
      ,@body
      (end-texture-mode)))
 
+;; FIXME Problem with this approach: entering a nested do-game-loop will overwrite *SCENE*
 (defmacro do-game-loop ((&key
+                           (scene nil)
                            (livesupport nil)
                            (vars ())
                            (end ())
                            (result ()))
                         &body body)
-  `(do ,vars ((or (window-should-close-p) ,end) ,result)
-     ,@(when livesupport `((declare (notinline))))
-     ,(if livesupport
-          `(livesupport:continuable
-             ,@body
-             (livesupport:update-repl-link))
-          `(progn ,@body))))
+  "Execute a game loop.
+
+When given, this will load SCENE, enable LIVESUPPORT during execution of the loop, expose the
+bindings in VARS to the loop BODY, stop the loop when END is non-nil, and return RESULT.
+
+To switch scenes in the loop body, use (SWITCH-SCENE MY-NEW-SCENE). SWITCH-SCENE loads the new scene
+and unloads the previous scene automatically. The current scene is accesible via *SCENE*."
+  `(progn
+     ;; Setup the initial *SCENE* as given by SCENE
+     (when ,scene
+       (setf *scene* ,scene)
+       (setup-scene *scene*))
+     (do ,vars ((or (window-should-close-p) ,end)
+                (tear-down-scene *scene*) ; Tear-down and reset *SCENE*
+                (setf *scene* nil)        ; at the end of the loop
+                ,result)
```

Comment:
Did this compile without warnings or errors? I believe your `do` form is invalid. Should require a `progn` around the result forms.

Reply:
Also, if `*scene*` is special then you won't need the `setf`.

Reply:
> Did this compile without warnings or errors?

Yes! Fun fact that I learned while writing this: `do` can take any number of `EXIT-FORM`s after the `TEST`, returning the value of the last one. So it's already implicitly a `progn`. Documentation: `DO ({(Var [Init] [Step])}*) (Test Exit-Form*) Declaration* Form*`